src/api/app/models/worker_status.rb

```diff
@@ -1,25 +1,46 @@
 class WorkerStatus
-  def self.hidden
-    mydata = Rails.cache.read('workerstatus')
-    ws = Nokogiri::XML(mydata || Backend::Api::BuildResults::Worker.status).root
-    prjs = {}
-    ws.css('building').each do |b|
-      prjs[b['project']] = 1
+  class << self
+    def hidden
+      mydata = Rails.cache.fetch('workerstatus') do
```

Comment:
Any reason to change from a `Rails.cache.read` call to a `Rails.cache.fetch` one?

Reply:
Yes unless @coolo says differently, I don't see a point to don't put the information in cache, using `fetch`.

Reply:
hmm, my knowledge about Rails.cache is not uptodate - but doesn't using `.fetch` implicitly write without expiry? Before it was reading and if the data is expired it reads from backend.

Having outdated cache data in /build/_workerstatus isn't really the point.

Reply:
> hmm, my knowledge about Rails.cache is not uptodate - but doesn't using `.fetch` implicitly write without expiry? Before it was reading and if the data is expired it reads from backend.
> 
> Having outdated cache data in /build/_workerstatus isn't really the point.

Hey @coolo thank you for the quick analysis.

I wondering if we are really going to have an outdated cache with something like that being executed by clockworkd:

```
  every(30.seconds, 'status.refresh') do
    # this should be fast, so don't delay
    WorkerStatus.new.update_workerstatus_cache
  end
```

I think it would be better to have the fetch expiring the key every 30 seconds (!!) and have just the `StatusHistory` generation running out of the main thread with a delayed job? 

In the clockwork which overwrites the same cache entry.

Reply:
but that writes into the cache - and that's the reason `hidden` is only reading. So it's always clear what expiry date applies?

Reply:
> hmm, my knowledge about Rails.cache is not uptodate - but doesn't using `.fetch` implicitly write without expiry?

unless you are setting the `expire_in`, as we do

Reply:
> but that writes into the cache - and that's the reason `hidden` is only reading. So it's always clear what expiry date applies?

I just think it's a waste of resources keep generating a StatusHistory and caching this information, every 30 seconds. I understand that StatusHistory is necessary to show some statistics (probably 30 seconds resolution is an overkill) but definitely if nobody is accessing the site you don't need to keep invalidating the caching every 30 seconds and asking the backend for new data...

Reply:
well, the /monitor graph can show an hour - so 120 points don't appear overkill to me.

Reply:
@eduardoj I rolled back this change, please review it again

Reply:
> well, the /monitor graph can show an hour - so 120 points don't appear overkill to me.

The meaning of the word overkill has nothing to do with what the graph is capable to show @coolo. Overkill is generate an excessive amount of data if it's not necessary to any business critical decision or actively invalidate the cache every 30 seconds regardless of user access or not, to save some milliseconds in the first hit in the monitor page. These whole statistics monitoring things should be done at https://obs-measure.opensuse.org/ anyway..

Reply:
If build.opensuse.org was the only instance, I'd agree - but it's not. And requiring a grafana installation to see how the scheduler backlog is sounds a little harsh.

And this is not to save time for hitting /monitor - this is about showing a graph about the last hour's events. Grafana light.